### PR TITLE
feat(api): Analyze v6 json protocols and translate v6 commands to protocol engine 

### PR DESCRIPTION
**Pull request by TamarZanzouri**
# Overview

Analyze a Json v6 protocol in the protocol engine

# Changelog

- Implemented _load_json methon in protocol_runner.py to accept a Json v6 protocol
- Implemented a translation layer from Json v6 command to the protocol engine commands

# Review requests

Please take a look at test_json_command_translator.py INVALID_TEST_PARAMS. Feels redundancy or not implemented correctly. Would love to have feedback!

# Risk assessment

Make sure this works on a robot and legacy protocols work as well

### CODE REVIEW

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 1]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
...
-        for labware_id, labware in protocol.labware.items():
-            result.append(
-                self._translate_load_labware(
-                    labware_id, labware, protocol.labwareDefinitions
-                )
-            )
-
+        """Takes json protocol v6 and translates commands->protocol engine commands."""
+        commands_list: List[pe_commands.CommandCreate] = []
+        exclude_commands = [
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py:107]: Can we link to the #9553 epic for these exclusions?

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 2]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
...
+                assert module_id is not None
+                assert modules is not None
+                dict_command["params"].update({"model": modules[module_id].model})
+            elif command.commandType == "loadLabware":
+                labware_id = command.params.labwareId
+                assert labware_id is not None
+                labware_command = _get_labware_command(protocol, labware_id, command)
+                commands_list.append(labware_command)
+                continue
+                print(dict_command)
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: This print should be removed

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 3]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
...
+                modules = protocol.modules
+                assert module_id is not None
+                assert modules is not None
+                dict_command["params"].update({"model": modules[module_id].model})
+            elif command.commandType == "loadLabware":
+                labware_id = command.params.labwareId
+                assert labware_id is not None
+                labware_command = _get_labware_command(protocol, labware_id, command)
+                commands_list.append(labware_command)
+                continue
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Why does the labware command logic behave differently than the pipette and module logic? I think we should pick one of the two approaches and use it consistently

1. Modify command params, pass it to regular translation (used by loadPipette and loadModule)
2. Construct a new command and add it to the list, bypassing regular translation (used by loadLabware)

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 4]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
...
-                pipetteId=command.params.pipette,
-                labwareId=command.params.labware,
-                wellName=command.params.well,
-                volume=command.params.volume,
-                wellLocation=WellLocation(
-                    origin=WellOrigin.BOTTOM,
-                    offset=WellOffset(x=0, y=0, z=command.params.offsetFromBottomMm),
+            dict_command = command.dict(exclude_none=True)
+            if command.commandType in exclude_commands:
+                continue
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: I find `continue` to be hard to follow in general. Can this be refactored to avoid `continue`?

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 5]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
...
+            if command.commandType == "loadPipette":
+                pipette_id = command.params.pipetteId
+                assert pipette_id is not None
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Can we add TODO's on these asserts that explain how they're needed because out Pydantic model for JSONv6 isn't strict enough to differentiate between the different command payload types?

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 6]
```diff
@@ -19,287 +11,73 @@ class CommandTranslatorError(Exception):
     pass
 
 
+def _get_labware_command(
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Do our tests cover this logic? Codecov doesn't seem to think so

File: api/src/opentrons/protocol_runner/json_file_reader.py [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+from opentrons_shared_data.protocol.models.protocol_schema_v6 import ProtocolSchemaV6
 from opentrons.protocol_reader import ProtocolSource
 
 
 class JsonFileReader:
     """Reads and parses JSON protocol files."""
 
     @staticmethod
-    def read(protocol_source: ProtocolSource) -> JsonProtocol:
+    def read(protocol_source: ProtocolSource) -> ProtocolSchemaV6:
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/json_file_reader.py:11]: We should add smoke tests for this using the fixtures

File: shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py [hunk 8]
```diff
@@ -152,3 +152,7 @@ class ProtocolSchemaV6(BaseModel):
     liquids: Optional[Dict[str, Liquid]]
     commandAnnotations: Optional[List[CommandAnnotation]]
     designerApplication: Optional[DesignerApplication]
+
+    class Config:
+        # added for constructing the class with field name instead of alias
+        allow_population_by_field_name = True
```
Comment (mcous) [inline shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py:158]: Why did we add this?

**Review by mcous — CHANGES_REQUESTED** [review]
Looking pretty good! Little batch of comments from just reading the code, looks like there's still some debugging code to remove

File: shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py (hunk 8 above)
Comment (SyntaxColoring) [reply shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py:158]: If memory serves, then bizarrely, without this, this is an error:

```python
my_protocol = ProtocolSchemaV6(
    otSharedSchema="#/protocol/schemas/6",
    ...
)
```

Because Pydantic expects us to use the alias `$otSharedSchema` (which is impossible because of the dollar sign) instead of the field name `otSharedSchema`.

Comment (SyntaxColoring) [reply shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py:158]: In other words, from what I can tell, Pydantic has it totally backwards.

* If you do `model.dict()` or `model.json()`, you're probably trying to serialize a Python object into JSON. So you probably want to use field aliases. But this isn't the default; you have to do `.model.json(by_alias=True)`.
* If you do `model = Model(some_field=123, ...)`, you're constructing a Python object from other Python objects. So you probably want to use Python field names. But this isn't the default; you have to do `allow_population_by_field_name=True`.

🙃

File: api/src/opentrons/protocol_runner/json_command_translator.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: I personally favor (2).

Comment (TamarZanzouri) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: totally agree! I will change them all to the second approach

File: api/src/opentrons/protocol_runner/json_command_translator.py (hunk 6 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: just added a test for this! Thank you!

File: shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py (hunk 8 above)
Comment (mcous) [reply shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py:158]: 🤦

File: robot-server/tests/integration/protocols/simpleV6.json [hunk 9]
```diff
@@ -0,0 +1,1519 @@
+{
```
Comment (mcous) [inline robot-server/tests/integration/protocols/simpleV6.json]: Can we symlink this in from `shared-data` instead of duplicating it?

File: robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml [hunk 10]
```diff
@@ -0,0 +1,1844 @@
...
+                    - D12
+                    - E12
+                    - F12
+                    - G12
+                    - H12
+                    metadata: {}
+              startedAt: '2022-03-24T20:20:13.583824+00:00'
+              completedAt: '2022-03-24T20:20:13.583887+00:00'
+            - id: 0dcf39da-ccf3-439f-b469-0b248e62cbce
+              createdAt: '2022-03-24T20:20:13.507775+00:00'
```
Comment (mcous) [inline robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml]: `id`, `createdAt`, `key` (for now) will not be stable across different analysis runs, nor are they intended to be. I don't think this test as written can pass

**Review by mcous — COMMENTED** [review]
Comments about the Tavern tests

File: api/src/opentrons/protocol_runner/json_command_translator.py (hunk 3 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: The prob I have with the second approach is that we are losing all auto mapping capabilities. If I convert it all to the second approach I need to map each field manually. @mcous @SyntaxColoring

File: robot-server/tests/integration/protocols/simpleV6.json (hunk 9 above)
Comment (TamarZanzouri) [reply robot-server/tests/integration/protocols/simpleV6.json]: @y3rsh we spoke about this! @mcous we wanted to import it but we can only import from the python folder?

File: robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml (hunk 10 above)
Comment (TamarZanzouri) [reply robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml]: I removed their value to support any string! thank you!

File: robot-server/tests/integration/protocols/simpleV6.json (hunk 9 above)
Comment (mcous) [reply robot-server/tests/integration/protocols/simpleV6.json]: I was suggesting creating a symbolic link via `ln -s` rather than copying the file. For example:

```shell
cd robot-server/tests/integration/protocols
ln -sf ../../../../shared-data/protocol/fixtures/6/simpleV6.json simpleV6.json
```

But it turns out, I think **this is unnecessary**. This test yaml worked fine for me, no symbolic link or file duplication required:

```yaml
stages:
  - name: Upload simple v6 protocol
    request:
      url: '{host:s}:{port:d}/protocols'
      method: POST
      files:
        files: '../shared-data/protocol/fixtures/6/simpleV6.json'
```

File: robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml [hunk 11]
```diff
@@ -0,0 +1,1844 @@
...
+marks:
+  - usefixtures:
+      - run_server
+stages:
+  - name: Upload simple v6 protocol
+    request:
+      url: '{host:s}:{port:d}/protocols'
+      method: POST
+      files:
+        files: 'tests/integration/protocols/simpleV6.json'
```
Comment (mcous) [inline robot-server/tests/integration/http_api/protocols/test_v6_json_protocol.tavern.yaml]: Let's not duplicate the fixture file unnecessarily

```suggestion
        files: '../shared-data/protocol/fixtures/6/simpleV6.json'
```

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 12]
```diff
@@ -19,287 +18,110 @@ class CommandTranslatorError(Exception):
...
-        pipette_id: str,
-        pipette: models.json_protocol.Pipettes,
-    ) -> pe_commands.LoadPipetteCreate:
-        return pe_commands.LoadPipetteCreate(
-            params=pe_commands.LoadPipetteParams(
-                pipetteName=PipetteName(pipette.name),
-                mount=MountType(pipette.mount),
-                pipetteId=pipette_id,
-            )
+def _get_labware_command(
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Naming nitpick, feel free to ignore for this PR:

I might have called these `_translate_labware_command()`, etc. instead of `_get_labware_command()`, etc.

"Get," to me, implies retrieving something that already exists, rather than creating something.

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 13]
```diff
@@ -19,287 +18,110 @@ class CommandTranslatorError(Exception):
...
-                wellLocation=WellLocation(
-                    origin=WellOrigin.BOTTOM,
-                    offset=WellOffset(x=0, y=0, z=command.params.offsetFromBottomMm),
-                ),
-            )
+    )
+    return labware_command
+
+
+def _get_command(command: Command) -> pe_commands.CommandCreate:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Naming nitpick, feel free to ignore for this PR:

I might rename this to `_get_other_command()` (or `_translate_other_command()`—see my other comment) and move it below the other 3 helper functions.

File: api/src/opentrons/protocol_runner/json_command_translator.py [hunk 14]
```diff
@@ -19,287 +18,110 @@ class CommandTranslatorError(Exception):
...
-                pipetteName=PipetteName(pipette.name),
-                mount=MountType(pipette.mount),
-                pipetteId=pipette_id,
-            )
+def _get_labware_command(
+    protocol: ProtocolSchemaV6, command: Command
+) -> pe_commands.LoadLabwareCreate:
+    labware_id = command.params.labwareId
+    # asserting labware_id and definition_id to raise an
+    # error in case they do not exist in the v6 model
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_runner/json_command_translator.py]: Nitpick, feel free to ignore for this PR:

I think the fact that this raises an error if these fields don't exist in the v6 model is self-evident from the code, and this comment should instead explain more about why we need to do this.

Namely: by JSON v6 specification, `command` *must* contain these fields. But, because of limitations in how we've modeled `params`, the type-checker doesn't know that; it thinks the fields might be `None`. And also, nothing before this `assert` has actually validated that the client provided a file that conforms to the spec in this way.

File: api/tests/opentrons/protocol_runner/test_json_command_translator.py [hunk 15]
```diff
@@ -13,170 +24,167 @@
...
+                labwareId="labware-id-def456",
+                wellName="A1",
+                wellLocation=WellLocation(),
+            )
+        ),
+    ),
+    (
+        json_v6_models.Command(
+            id="load-pipette-command-id-666",
+            commandType="loadPipette",  # used to be delay but is expecting pause
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_runner/test_json_command_translator.py]: There's a few of these comments on commands unrelated to pausing/delaying. Maybe they're left over from a copy-paste?

File: api/tests/opentrons/protocol_runner/test_json_command_translator.py [hunk 16]
```diff
@@ -218,71 +223,204 @@ def test_dispense(subject: JsonCommandTranslator) -> None:
...
+
+
+@pytest.mark.parametrize("test_input, expected_output", INVALID_TEST_PARAMS)
+def test_invalid_commands(
+    subject: JsonCommandTranslator,
+    test_input: json_v6_models.Command,
+    expected_output: pe_commands.CommandCreate,
+) -> None:
+    """Test fail invalid payload-translating v6 commands to protocol engine commands."""
+    with pytest.raises(AssertionError):
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_runner/test_json_command_translator.py:423]: I think this is fine for now, but in general:

1. If a test tests that a function raises something, it's *usually* because the fact that it raises that thing is part of the function's public, documented interface. (Otherwise, why would we be testing it?)
2. If a function raises something as part of its interface, it should do `raise MyException()` instead of `assert`. `assert` statements [get disabled with optimizations](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement). That makes them good for debugging, but not good for reporting actual cases where the function is miscalled.

I'd argue that this holds true for `JsonCommandTranslator`. It's given untrusted input, and part of its job is to flag when that input is wrong, instead of letting the problem propagate forward. But since we know [we shouldn't have to do this, anyway](https://github.com/Opentrons/opentrons/issues/9803), I'm fine leaving it as-is in this PR, if we don't want to go to the trouble of making a new exception type to raise.

File: shared-data/python/opentrons_shared_data/protocol/models/__init__.py [hunk 17]
```diff
@@ -1,8 +1,30 @@
...
+    "Labware",
+    "Command",
+    "Pipette",
+    "Metadata",
+    "Params",
+    "OffsetVector",
+    "WellLocation",
+    "Module",
+    "Location",
 ]
```
Comment (SyntaxColoring) [inline shared-data/python/opentrons_shared_data/protocol/models/__init__.py]: No action needed for this PR, just thinking aloud:

I think re-exporting all these models indiscriminately is going to make things a bit confusing for calling code. What does `opentrons_shared_data.protocol.models.Robot` mean? It is, in fact, a `Robot` object as seen in JSON schema v6—but the import path makes it seem much broader than that.

Maybe we should reorganize things like:

* `opentrons_shared_data.protocol.models.v6.Protocol`
* `opentrons_shared_data.protocol.models.v6.Robot`
* etc.

Or even:

* `opentrons_shared_data.protocol.v6.Protocol`
* `opentrons_shared_data.protocol.v6.Robot`
* etc.

**Review by SyntaxColoring — COMMENTED** [review]
Some final comments, but otherwise looks good to me!

-----

I can't leave a comment on the file because it's empty, but It looks like the `test.txt` file might have been committed accidentally?

File: shared-data/python/opentrons_shared_data/protocol/models/__init__.py (hunk 17 above)
Comment (TamarZanzouri) [reply shared-data/python/opentrons_shared_data/protocol/models/__init__.py]: I can change the name to 'MachineDetails' or 'RobotDetails'. I agree this name is expected to have much broader details then it actually has.

File: api/src/opentrons/protocol_runner/json_command_translator.py (hunk 13 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: I changed this to _translate_simple_command()

File: api/src/opentrons/protocol_runner/json_command_translator.py (hunk 14 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_runner/json_command_translator.py]: I totally agree! I will change it!

File: shared-data/python/opentrons_shared_data/protocol/models/__init__.py (hunk 17 above)
Comment (SyntaxColoring) [reply shared-data/python/opentrons_shared_data/protocol/models/__init__.py]: I don't think that will help, because there are `MachineDetails` or `RobotDetails` for v5 protocols, too, which aren't necessarily the same model. In other words, I'm saying the class name `Robot` is perfectly fine—it's the *package name*, `models`, that's potentially ambiguous.

Comment (mcous) [inline shared-data/python/opentrons_shared_data/protocol/models/__init__.py]: It looks like most of these re-exports aren't being used, and I agree with Max's comment. I know we've got some weirdness still to sort out in `opentrons_shared_data.protocols`, so let's do this for now:

```suggestion
from . import protocol_schema_v6
from .protocol_schema_v6 import ProtocolSchemeV6

__all__ = [
    "ProtocolSchemaV6",
    "protocol_schema_v6",
]
```

Export the main model (`ProtocolSchemaV6`) at the top level, but also re-export the entire `protocol_schema_v6` module for explicitly public access to all the subtypes

**Review by mcous — COMMENTED** [review]
Looks like `robot-server/tests/integration/http_api/protocols/test.txt` still needs to be removed, and I agree with Max's comment on the module exports from shared data.

I think these are the last two things I see for this PR, so we should be good to go shortly

File: shared-data/python/opentrons_shared_data/protocol/models/__init__.py (hunk 17 above)
Comment (TamarZanzouri) [reply shared-data/python/opentrons_shared_data/protocol/models/__init__.py:30]: I can do that but don't we want to change the Robot class name anyway? @mcous it does feel too robust to what's in it

Comment (TamarZanzouri) [reply shared-data/python/opentrons_shared_data/protocol/models/__init__.py]: after reading this again I understand your point. thank you for pointing that

**Review by SyntaxColoring — APPROVED** [review]
Thanks for the changes.

It looks like CI is failing because of a formatting/linting problem. Once that's fixed and CI is passing, this looks good to me to merge. ✨

**Review by mcous — APPROVED** [review]
